packages/0xcert-ethereum-gateway-contracts/src/contracts/token-deploy-gateway.sol

```diff
@@ -0,0 +1,411 @@
+    uint256 seed;
+    uint256 expiration;
+  }
+
+  /**
+   * @dev Address of token transfer proxy.
+   */
+  address public tokenTransferProxy;
+
+  /**
+   * @dev Mapping of all cancelled deploys.
+   */
+  mapping(bytes32 => bool) public deployCancelled;
+
+  /**
+   * @dev Mapping of all performed deploys.
+   */
+  mapping(bytes32 => bool) public deployPerformed;
+
+  /**
+   * @dev This event emmits when deploy is performed.
+   */
+  event Perform(
+    address indexed _maker,
+    address indexed _taker,
+    address _createdContract,
+    bytes32 _claim
+  );
+
+  /**
+   * @dev This event emmits when deploy is cancelled.
+   */
+  event Cancel(
+    address indexed _maker,
+    address indexed _taker,
+    bytes32 _claim
+  );
+
+  /**
+   * @dev Constructor sets token transfer proxy address.
+   * @param _tokenTransferProxy Address of token transfer proxy.
+   */
+  constructor(
+    address _tokenTransferProxy
+  )
+    public
+  {
+    tokenTransferProxy = _tokenTransferProxy;
+  }
+  
+  /**
+   * @dev Performs the atomic swap that deploys a new token smart contract and at the same time
+   * transfers tokens.
+   * @param _data Data required to make the deploy.
+   * @param _signature Data from the signature.
+   */
+  function perform(
+    DeployData memory _data,
+    SignatureData memory _signature
+  )
+    public
+  {
+    require(_data.taker == msg.sender, TAKER_NOT_EQUAL_TO_SENDER);
+    require(_data.expiration >= now, CLAIM_EXPIRED);
+
+    bytes32 claim = getDeployDataClaim(_data);
+    require(
+      isValidSignature(
+        _data.maker,
+        claim,
+        _signature
+      ),
+      INVALID_SIGNATURE
+    );
+
+    require(!deployCancelled[claim], DEPLOY_CANCELED);
+    require(!deployPerformed[claim], DEPLOY_ALREADY_PERFORMED);
+
+    deployPerformed[claim] = true;
+
+    address token = _doActions(_data);
+
+    emit Perform(
+      _data.maker,
+      _data.taker,
+      token,
+      claim
+    );
+  }
+
+  /**
+   * @dev @dev Performs the atomic swap that deploys a new token smart contract and at the same time
+   * transfers tokens where performing address does not need to be known before
+   * hand.
+   * @notice When using this function, be aware that the zero address is reserved for replacement
+   * with msg.sender, meaning you cannot send anything to the zero address.
+   * @param _data Data required to make the deploy.
+   * @param _signature Data from the signature.
+   */
+  function performAnyTaker(
+    DeployData memory _data,
+    SignatureData memory _signature
+  )
+    public
+  {
+    require(_data.expiration >= now, CLAIM_EXPIRED);
+
+    bytes32 claim = getDeployDataClaim(_data);
+    require(
+      isValidSignature(
+        _data.maker,
+        claim,
+        _signature
+      ),
+      INVALID_SIGNATURE
+    );
+
+    require(!deployCancelled[claim], DEPLOY_CANCELED);
+    require(!deployPerformed[claim], DEPLOY_ALREADY_PERFORMED);
+
+    deployPerformed[claim] = true;
```

Comment:
This is to prevent a reentry attack. Since in _doActions we call an external smart contract that could be malicious and try to call token deploy gateway again it would fail since the deploy is already marked as performed.